### ISSUE TITLE
Question about _get_batch_logps of dpo_trainer.py

### ISSUE THREAD

**Issue by wulaoshi** [huggingface/trl#1095]
```
def _get_batch_logps(
        self,
        logits: torch.FloatTensor,
        labels: torch.LongTensor,
        average_log_prob: bool = False,
    ) -> torch.FloatTensor:
        if logits.shape[:-1] != labels.shape:
            raise ValueError("Logits (batch and sequence length dim) and labels must have the same shape.")

        if not self.is_encoder_decoder:
            labels = labels[:, 1:].clone()
            logits = logits[:, :-1, :]
        loss_mask = labels != self.label_pad_token_id

        # dummy token; we'll ignore the losses on these tokens later
        labels[labels == self.label_pad_token_id] = 0

        per_token_logps = torch.gather(logits.log_softmax(-1), dim=2, index=labels.unsqueeze(2)).squeeze(2)

        if average_log_prob:        #Why is a sum(-1) operation needed here?
            return (per_token_logps * loss_mask).sum(-1) / loss_mask.sum(-1)
        else:
            return (per_token_logps * loss_mask).sum(-1)
```` 

First of all, I'd like to express my gratitude for the amazing work done by the author. While going through the DPO code, I came across a point of confusion: I understand that π(y|x) represents the probability distribution when the model generates a response, but in the code implementation, there's an additional sum(-1) operation. However, I don't see summation operation in the formula, as shown in the image below:
<img width="767" alt="image" src="https://github.com/huggingface/trl/assets/27938964/65ba3fa1-50a9-42b6-8d21-ec7983806ca2">

At the same time, the sum operation is again different from the π(y|x) implementation in PPO:
```
def logprobs_from_logits(logits, labels, gather=True):
    logp = F.log_softmax(logits, dim=2)

    if not gather:
        return logp
    logpy = torch.gather(logp, 2, labels.unsqueeze(2)).squeeze(-1)
    return logpy     # No sum(-1)
```

Could you please help me understand the logic behind this implementation? Thank you!

**Comment by kashif** [comment]
@wulaoshi  that part of the code is not used as `average_log_prob=False` I believe the original authors had it for development... but we should remove it in the TRL implementation

**Comment by kashif** [comment]
@wulaoshi the `sum(-1)` is due to the log, since by the chain rule of probability the logits become a sum when doing logs... over each generated token

**Comment by wulaoshi** [comment]
> @wulaoshi the `sum(-1)` is due to the log, since by the chain rule of probability the logits become a sum when doing logs... over each generated token

I think I understand now: in the _get_batch_logps method in DPO, π(y|x) refers to the generation probability of sentence y, so it is the product of the probabilities of each token, which eventually translates to sum(-1). In PPO, logprobs_from_logits is used to obtain the probability distribution of sentence y to facilitate the calculation of the KL distance later. Is my understanding correct? Thank you for your response.

**Comment by kashif** [comment]
yes

**Comment by wulaoshi** [comment]
> yes

Thank you once again for your response. To better illustrate my question, I have also included the formula from the paper in my previous reply.
